lib/BuildSystem/ExternalCommand.cpp

```diff
@@ -242,19 +244,30 @@ void ExternalCommand::provideValue(BuildSystemCommandInterface& bsci,
     // FIXME: Eventually, it might be nice to harden the format so that we know
     // when an output was actually required versus optional.
     if (value.isMissingOutput())
-      return false;
+      return llvm::None;
 
     // If the value is a missing input, but those are allowed, it is ok.
-    if (allowMissingInputs && value.isMissingInput())
-      return false;
+    if (value.isMissingInput()) {
+      if (allowMissingInputs)
+        return llvm::None;
+      else
+        return BuildValue::makeMissingInput();
+    }
 
-    // For anything else, this is an error and the command should be skipped.
-    return true;
+    // Propagate cancellation.
+    if (value.isCancelledInput())
+      return BuildValue::makeCancelledCommand();
```

Comment:
This doesn't seem right to me -- a command which is never started isn't cancelled, it is the "propagated failure" case.